# 2. teden: Numpy - Domača naloga

Uvozimo modul numpy:

In [1]:
import numpy as np

### 1.Create a null vector of size 10

In [3]:
np.zeros(10)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

Odg: `[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]`

### 2.Create a null vector of size 10 but the fifth value which is 1

In [4]:
a=np.zeros(10)
a[4]=1
a

array([0., 0., 0., 0., 1., 0., 0., 0., 0., 0.])

Odg: `[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]`

### 3.Create a vector with values ranging from 10 to 49

In [7]:
np.linspace(10,49,40)

array([10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22.,
       23., 24., 25., 26., 27., 28., 29., 30., 31., 32., 33., 34., 35.,
       36., 37., 38., 39., 40., 41., 42., 43., 44., 45., 46., 47., 48.,
       49.])

Odg: `[10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]`

### 4.Reverse a vector (first element becomes last)

In [10]:
Z = np.arange(50)

In [14]:
Z

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49])

In [15]:
Z[::-1]

array([49, 48, 47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33,
       32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16,
       15, 14, 13, 12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0])

Odg: `[49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28 27 26 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2 1  0]`

### 5.Create a 3x3 matrix with values ranging from 0 to 8

**Pomoč**: metoda .reshape() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html)

In [18]:
a=np.linspace(0,8,9).reshape(3,3)
a

array([[0., 1., 2.],
       [3., 4., 5.],
       [6., 7., 8.]])

Odg: 

`[[0 1 2]
 [3 4 5]
 [6 7 8]]`

### 6.Find indices of non-zero elements from [1,2,0,0,4,0]

**Pomoč:** Google, Stack Overflow, Numpy dokumentacija

In [24]:
a = [1,2,0,0,4,0]

In [25]:
np.nonzero(a)

(array([0, 1, 4], dtype=int64),)

Odg: `(array([0, 1, 4], dtype=int64),)`

### 7.Create a 3x3 identity matrix

In [26]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

Odg: 

`[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]`

### 8.Create a 3x3x3 array with random values

In [30]:
np.random.rand(3,3,3)

array([[[0.03696858, 0.04159908, 0.85620071],
        [0.71659035, 0.35383425, 0.80312566],
        [0.46216653, 0.8908113 , 0.43117705]],

       [[0.18951397, 0.49273487, 0.24785337],
        [0.28007546, 0.54190244, 0.09436871],
        [0.48117264, 0.71189673, 0.97697233]],

       [[0.03672755, 0.93783078, 0.16082351],
        [0.19690611, 0.45695323, 0.91118359],
        [0.31876459, 0.79979493, 0.01123951]]])

Odgovor npr.: 
    
`[[[0.61712056 0.38943629 0.16277319]
  [0.19481511 0.12725415 0.6717974 ]
  [0.15340477 0.78713636 0.52864134]]

 [[0.91153827 0.7756351  0.88600077]
  [0.03509814 0.79676837 0.39167056]
  [0.55410251 0.67583417 0.64169036]]

 [[0.46862042 0.05954143 0.42512482]
  [0.34746791 0.97653267 0.0024134 ]
  [0.53522577 0.37192358 0.29601563]]]`

### 9.Create a 10x10 array with random values and find the minimum and maximum values

In [32]:
ra=np.random.rand(10,10)
ra

array([[0.17779062, 0.66656439, 0.46465425, 0.08980539, 0.85418077,
        0.30496637, 0.77507841, 0.14053318, 0.46739735, 0.30444051],
       [0.00896928, 0.3305398 , 0.79157065, 0.71875282, 0.95277442,
        0.77924619, 0.96336214, 0.68325448, 0.76162707, 0.76109687],
       [0.14189521, 0.98766625, 0.69937117, 0.21793536, 0.58189964,
        0.6250665 , 0.26105619, 0.41438663, 0.1566239 , 0.19963972],
       [0.54645194, 0.59563286, 0.74564226, 0.12001012, 0.35936959,
        0.46635334, 0.46736808, 0.00668062, 0.64232904, 0.42421663],
       [0.76097053, 0.80575792, 0.72587143, 0.18161493, 0.97134177,
        0.22214429, 0.49078937, 0.47704181, 0.30207431, 0.33893192],
       [0.01330282, 0.54948378, 0.43932903, 0.85701139, 0.12488678,
        0.38215176, 0.56443836, 0.98872326, 0.84077448, 0.97147463],
       [0.3943478 , 0.45118135, 0.92131702, 0.27339091, 0.89541817,
        0.28190571, 0.88708851, 0.2323654 , 0.32679077, 0.10071018],
       [0.83765014, 0.16826694, 0.7510896

In [35]:
Zmax=ra.max()
Zmax

0.98872325641425

In [36]:
Zmin=ra.min()
Zmin

0.0066806246708722705

Odg: `Zmin= 0.004874889654358006 Zmax= 0.992254173881756`

### 10.Create a 2d array (10x10) with 1 on the border and 0 inside

In [41]:
a=np.ones([10,10])
a

array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

In [42]:
a[1:9,1:9]=0
a

array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

Odg: 
    
`[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]`

### 11.What is the result of the following expression?

In [43]:
0 * np.nan

nan

In [44]:
np.nan == np.nan

False

In [45]:
np.inf > np.nan

False

In [46]:
np.nan - np.nan

nan

In [48]:
0.3 == 3 * 0.1 

False

### 12.Create a 8x8 matrix and fill it with a checkerboard pattern

In [69]:
a=np.ones([8,8])
a[::2]=0
a=a.T
a[::2]=a[::2]-1
a=a+1
a[a==2]=0

In [70]:
a

array([[0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.]])

Odg: 

`[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]`

### 13.Given a 1D array, negate all elements which are between 3 and 8, in place.

In [71]:
Z = np.arange(11)

In [72]:
Z

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [73]:
Z[(Z>3)&(Z<=8)]=-Z[(Z>3)&(Z<=8)]

In [74]:
Z

array([ 0,  1,  2,  3, -4, -5, -6, -7, -8,  9, 10])

Odg: `[ 0  1  2  3 -4 -5 -6 -7 -8  9 10]`

### 14.Create a vector of size 10 with values ranging from 0 to 1, both excluded

In [77]:
np.linspace(0,1,11,endpoint=False)[1:]

array([0.09090909, 0.18181818, 0.27272727, 0.36363636, 0.45454545,
       0.54545455, 0.63636364, 0.72727273, 0.81818182, 0.90909091])

Odg: `[0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455 0.63636364 0.72727273 0.81818182 0.90909091]`

### 15.Create a random vector of size 10 and sort it

In [85]:
ra=np.random.rand(10)
ra

array([0.15178037, 0.55903904, 0.34508487, 0.08989044, 0.89099133,
       0.29730781, 0.92721772, 0.71624112, 0.88137231, 0.14561146])

In [86]:
ra=np.sort(ra)
ra

array([0.08989044, 0.14561146, 0.15178037, 0.29730781, 0.34508487,
       0.55903904, 0.71624112, 0.88137231, 0.89099133, 0.92721772])

Odg. npr.: `[0.08755811 0.17786018 0.45041482 0.48746541 0.68379823 0.70209576 0.77273884 0.86106446 0.93083351 0.98349603]`

### 16. Consider two random array A anb B, check if they are equal

In [94]:
A = np.random.randint(0,2,5)
A

array([0, 0, 0, 0, 1])

In [90]:
B = np.random.randint(0,2,5)
B

array([0, 0, 0, 0, 1])

In [95]:
min(A==B)

True

### 17.Make an array immutable (read-only)

In [99]:
Z = np.zeros(10)
Z

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [100]:
Z.flags.writeable = False

In [101]:
Z[1]=1

ValueError: assignment destination is read-only

### 18.Create random vector of size 10 and replace the maximum value by 0

In [102]:
ra=np.random.rand(10)
ra

array([0.24622941, 0.76771194, 0.81467358, 0.73336814, 0.28359736,
       0.61790591, 0.35445647, 0.89227275, 0.98056552, 0.68002985])

In [104]:
ra[ra==ra.max()]=0
ra

array([0.24622941, 0.76771194, 0.81467358, 0.73336814, 0.28359736,
       0.61790591, 0.35445647, 0.89227275, 0.        , 0.68002985])

### 19.How to find the closest value (to a given scalar -v) in an array -Z?

In [105]:
Z = np.arange(100)
Z

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

In [110]:
v = np.random.uniform(0,100)
v

32.551014014571024

In [111]:
ind=(np.abs(Z-v)).argmin()
ind

33

In [113]:
Z[ind]

33

### 20.How to convert an integer (32 bits)  array into a float (32 bits) in place?

In [114]:
Z = np.arange(10, dtype=np.int32)

In [115]:
print(Z)

[0 1 2 3 4 5 6 7 8 9]


In [116]:
Z=Z.astype('float32')

In [117]:
Z

array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.], dtype=float32)

Odg: `[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]`

### 21.How to find the most frequent value in an array?

**Pomoč:** metoda .bincount() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.bincount.html?highlight=bincount#numpy.bincount)

In [118]:
Z = np.random.randint(0,10,50)
print(Z)

[3 9 2 7 3 1 2 1 6 6 1 5 7 8 7 7 3 4 7 0 5 3 7 7 4 1 2 6 4 8 1 6 6 9 4 5 5
 7 8 4 7 4 8 1 1 7 2 0 9 6]


In [146]:
a=np.bincount(Z)
a

array([ 2,  7,  4,  4,  6,  4,  6, 10,  4,  3], dtype=int64)

In [147]:
np.argmax(a)

7